src/optimizer/topn_optimizer.cpp

```diff
@@ -12,6 +12,13 @@ bool TopN::CanOptimize(LogicalOperator &op) {
 	    op.children[0]->type == LogicalOperatorType::LOGICAL_ORDER_BY) {
 		auto &limit = op.Cast<LogicalLimit>();
 
+		// When there are some expressions in the limit operator,
+		// we shouldn't use this optimizations. Because of the expressions
+		// will be lost when it convert to TopN operator.
+		if (limit.limit || limit.offset) {
```

Comment:
@Tmonster Thanks for your review. I'll give you some contexts here. https://github.com/duckdb/duckdb/pull/1355 This PR support the non-constant value in limit clause, the expression will be stored in the `limit.limit` and `limit.offset`. But in the [`LogicalTopN`](https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/planner/operator/logical_top_n.hpp), it doesn't store the expression in the `limit.limit` and `limit.offset`. So those expression will be lost in [here](https://github.com/duckdb/duckdb/blob/master/src/optimizer/topn_optimizer.cpp#L29). So my method here is to forbid the topn optimization when we find some expressions in the limit clause. And I think another method is to support the expression in the TopN operator. But I think we can forbid this optimization first and add the expression into the TopN operator it later.

Reply:
I think I figured it out. A lot of my confusion comes from the fact that you have a limit.limit_val and limit.offset_val, and that`limit.limit_val =  NumericLimits<int64_t>::Maximum()` when there is no limit expression, and that `limit.offset_val = 0` when there is no offset expression. 

I couldn't manage to break it, but it is strange that a query like  `SELECT * FROM t ORDER BY x limit 5 OFFSET (SELECT 5);` doesn't apply the optimization. The (select 5) can be eagerly evaluated in the optimizer and replaced with the value 5, but it seems like the optimizer never does this? 
@pdet Do you know of any reason why this doesn't happen? Or is it an improvement that we can make later

Reply:
The reason is because [here](https://github.com/duckdb/duckdb/blob/master/src/planner/binder/query_node/plan_subquery.cpp#L102-L104). It use the cross join to handle the scalar subquery. But I have the same feeling like you, it can be treat as a single value.